### PR TITLE
New Adapter: Сointraffic

### PR DESCRIPTION
**Pull request by sergeykcointraffic**
Adds Prebid Server support.

## 🏷 Type of documentation
- [x] new bid adapter
- [ ] update bid adapter
- [ ] new feature
- [ ] text edit only (wording, typos)
- [ ] bugfix (code examples)
- [ ] new examples

## 🏷 Description of change
- [x] official adapter submission

- Contact email of the adapter's maintainer tech@cointraffic.io
- A link to a PR on the docs repo at prebid/prebid.github.io#5323

### CODE REVIEW

File: adapters/cointraffic/cointraffic.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	"github.com/prebid/prebid-server/v2/errortypes"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+	"net/http"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+// Builder builds a new instance of the Foo adapter for the given bidder with the given config.
```
Comment (SyntaxNode) [inline adapters/cointraffic/cointraffic.go:18]: Thanks for using the template from the docs. Please rename "Foo" to "Cointraffic" for your adapter.

File: openrtb_ext/imp_cointraffic.go [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package openrtb_ext
+
+type ExtImpCointraffic struct {
+	PlacementId string `json:"PlacementId"`
```
Comment (SyntaxNode) [inline openrtb_ext/imp_cointraffic.go:4]: Please consider using all lower case "placementid" or snake case "placementId" for bidder parameters.

File: adapters/cointraffic/cointraffic.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "No impression in the bid request",
+		}}
+	}
```
Comment (onkarvhanumante) [inline adapters/cointraffic/cointraffic.go:31]: should exclude this check from adapter PR. Prebid server core framework checks for impression length before invoking adapter code

File: adapters/cointraffic/cointraffic.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
```
Comment (onkarvhanumante) [inline adapters/cointraffic/cointraffic.go:79]: could make use of response utils
https://github.com/prebid/prebid-server/blob/ec729e6e5d50a1306d76332fda52108b9d313c1d/adapters/response.go#L10-L28

File: adapters/cointraffic/cointraffic.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(response.SeatBid[0].Bid))
+	bidResponse.Currency = response.Cur
```
Comment (onkarvhanumante) [inline adapters/cointraffic/cointraffic.go:86]: `adapters.NewBidderResponseWithBidsCapacity`  initialises `bidResponse.Currency` with `USD` value

consider scenario where `bidResponse.Currency`  is empty,  assigning `response.Cur` directly may set `bidResponse.Currency`  to empty value 

consider implementing following changes: 

```
	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(response.SeatBid[0].Bid))
        if response.Cur != "" {
	 bidResponse.Currency = response.Cur
       }
```

File: static/bidder-info/cointraffic.yaml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+endpoint: "https://apps.adsgravity.io/pbs/v1/request"
```
Comment (onkarvhanumante) [inline static/bidder-info/cointraffic.yaml:1]:
```
% curl -i --location --request POST 'https://apps.adsgravity.io/pbs/v1/request'
HTTP/2 400
date: Wed, 05 Jun 2024 11:16:14 GMT
content-type: text/javascript;charset=UTF-8
```

endpoint is reachable

File: static/bidder-info/cointraffic.yaml [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+maintainer:
+  email: tech@cointraffic.io
```
Comment (onkarvhanumante) [inline static/bidder-info/cointraffic.yaml:3]: Prebid team has sent email to verify above maintainers email. Requesting to responding back on email thread

File: adapters/cointraffic/cointraffictest/exemplary/simple-banner.json [hunk 8]
```diff
@@ -0,0 +1,125 @@
+{
+  "mockBidRequest": {
+    "id": "testRequestIdSingle",
+    "site": {
+      "page": "https://publisher-website.com/web"
+    },
+    "imp": [
+      {
+        "id": "testImpressionId",
```
Comment (onkarvhanumante) [inline adapters/cointraffic/cointraffictest/exemplary/simple-banner.json:9]: should update json test where request has multiple imp

File: static/bidder-info/cointraffic.yaml (hunk 7 above)
Comment (gargcreation1992) [reply static/bidder-info/cointraffic.yaml:3]: @sergeykcointraffic Please reply to email sent from us.

Comment (gargcreation1992) [reply static/bidder-info/cointraffic.yaml:3]: Screenshot 2024-06-24 at 7 13 46 PM

We have not received it yet. Please send it again. Is the email bouncing?

Comment (stsepelin) [reply static/bidder-info/cointraffic.yaml:3]: @gargcreation1992 Replied.

File: adapters/cointraffic/cointraffic.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	requestData, err := a.buildRequest(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func (a *adapter) buildRequest(request *openrtb2.BidRequest) (*adapters.RequestData, error) {
```
Comment (onkarvhanumante) [inline adapters/cointraffic/cointraffic.go:33]: `buildRequest` only builds headers and marshals request. This logic can be moved inside `MakeRequests` function

**Comment by onkarvhanumante** [comment]
> Hey @onkarvhanumante! Could you review this pull request again?

good to merge after addressing https://github.com/prebid/prebid-server/pull/3647#discussion_r1652710834

File: static/bidder-info/cointraffic.yaml (hunk 7 above)
Comment (gargcreation1992) [reply static/bidder-info/cointraffic.yaml:3]: @stsepelin We have not received it. Could you please check the outbox? Also please check if your email policies allows you to send email outside your org.

Comment (stsepelin) [reply static/bidder-info/cointraffic.yaml:3]: @gargcreation1992 Everything seems fine. Could you also check the spam folder/filters?

**Review by onkarvhanumante — DISMISSED** [review]

File: static/bidder-info/cointraffic.yaml (hunk 7 above)
Comment (bsardo) [reply static/bidder-info/cointraffic.yaml:3]: I see your responses in the spam folder.

Comment (stsepelin) [reply static/bidder-info/cointraffic.yaml:3]: I will check if there is anything we can do with this.

File: openrtb_ext/bidders.go [hunk 10]
```diff
@@ -80,6 +80,7 @@ var coreBidderNames []BidderName = []BidderName{
 	BidderCadentApertureMX,
 	BidderCcx,
 	BidderCoinzilla,
+	BidderCointraffic,
```
Comment (bsardo) [inline openrtb_ext/bidders.go:83]: Super nitpick: please move this above `BidderCoinzilla` to maintain alphabetical order.

File: openrtb_ext/bidders.go [hunk 11]
```diff
@@ -368,6 +369,7 @@ const (
 	BidderCadentApertureMX  BidderName = "cadent_aperture_mx"
 	BidderCcx               BidderName = "ccx"
 	BidderCoinzilla         BidderName = "coinzilla"
+	BidderCointraffic       BidderName = "cointraffic"
```
Comment (bsardo) [inline openrtb_ext/bidders.go:372]: Super nitpick: please move this above `BidderCoinzilla` to maintain alphabetical order.

**Review by bsardo — COMMENTED** [review]
This is looking good; I have just two nitpicks.

**Comment by stsepelin** [comment]
@bsardo Thanks for the review. Those two nitpicks are fixed now :)

**Review by onkarvhanumante — APPROVED** [review]

**Review by onkarvhanumante — APPROVED** [review]

**Review by bsardo — APPROVED** [review]

**Pull request merged**